File: BinarySearch/python/koko_eating_bananas.py
# Leet code problem no 875 

import math

def calucalteTimeTakenToEatEachBanana(files,no_of_bananas):
    total_hours=0
    for bananas in files:
        total_hours += math.ceil(bananas/no_of_bananas)

    return total_hours


def getTheMinimumNoOfHoursToEatAllBananas(files, minimum_hours):

    low = 1;
    high = max(files)

    while(low < high):
        mid = int(low + (high-low)/2)

        time_taken = calucalteTimeTakenToEatEachBanana(files,mid)

        if(time_taken <= minimum_hours):
            high = mid;
        else:
            low = mid +1

    return low



if __name__ == '__main__':
    N = 4,
    a = [3,6,7,11]
    h = 8
    result = getTheMinimumNoOfHoursToEatAllBananas(a,h)
    print(result)

File: BinarySearch/cpp/koko_eating_bananas.cpp
//
// Created by santosh on 12/31/25.
//


#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    long long calculateTimeTakenToEatEachBanana(vector<int>& piles, int bananas_per_hour) {
        long long total_hours = 0;

        for (int bananas : piles) {
            // ceil(bananas / bananas_per_hour) using integer math
            total_hours += (bananas + bananas_per_hour - 1) / bananas_per_hour;
        }

        return total_hours;
    }

    int getTheMinimumNoOfHoursToEatAllBananas(vector<int>& piles, int minimum_hours) {
        int low = 1;
        int high = *max_element(piles.begin(), piles.end());

        while (low < high) {
            int mid = low + (high - low) / 2;

            long long time_taken = calculateTimeTakenToEatEachBanana(piles, mid);

            if (time_taken <= minimum_hours) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }

        return low;
    }

    int minEatingSpeed(vector<int>& piles, int h) {
        return getTheMinimumNoOfHoursToEatAllBananas(piles, h);
    }
};


int main()
{
    Solution s;
    vector<int> piles={};
}